### PR TITLE
Add more configurable SSL options

### LINKED ISSUE #19: Expose client TLS certificate settings
Follow-up from [this Reddit thread](https://www.reddit.com/r/rust/comments/athkan/any_alternative_to_reqwest_in_rust/eh3yftj/?context=3).

----------

`libcurl` exposes settings for using TLS certificates presented by the client when initiating a TLS connection. This is used for so-called certificate-based authentication.

In `curl` (the CLI tool) these settings are controlled by the `--cert` / `--key` and `--cert-type` / `--key-type` flags. In the Rust bindings to `libcurl`, these correspond to [`Easy2::ssl_cert`](https://docs.rs/curl/0.5.0/curl/easy/struct.Easy2.html#method.ssl_cert), [`Easy2::ssl_key`](https://docs.rs/curl/0.5.0/curl/easy/struct.Easy2.html#method.ssl_key) and the `_type` methods.

It would be very useful to have those exposed in `cHTTP` in some way!

### PR DESCRIPTION
**Pull request by sagebind**
Provide request options for configuring a custom client certificate and private key for SSL/TLS connections.

Also expose the option for listing specific SSL/TLS cipher names to enable.

Resolves #19.

### CODE REVIEW

**Comment by sagebind** [comment]
@tazjin What do you think? Would these new properties on `Options` solve your use case?

File: src/options.rs [hunk 1]
```diff
@@ -84,15 +114,18 @@ impl Default for Options {
...
     }
 }
 
 /// Describes a policy for handling server redirects.
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
 pub enum RedirectPolicy {
-    /// Do not apply any special treatment to redirect responses. The response will be return as-is and redirects will
-    /// not be followed.
+    /// Do not apply any special treatment to redirect responses. The response
+    /// will be return as-is and redirects will not be followed.
```
Comment (tazjin) [inline src/options.rs]: > will be return**ed**

File: src/options.rs [hunk 2]
```diff
@@ -107,3 +140,70 @@ impl Default for RedirectPolicy {
...
+        }
+    }
+}
+
+/// A private key file.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PrivateKey {
+    pub(crate) path: PathBuf,
+    pub(crate) format: CertificateFormat,
+    pub(crate) password: String,
```
Comment (tazjin) [inline src/options.rs]: Not every private key file has a password. I think most PKCS12-encoded key pairs will, but at least with PEM formats its not unusual to have an unencrypted key.

**Comment by tazjin** [comment]
@sagebind I left two small comments. Looks good otherwise! I like the encoding of the key types in the type system :+1:

**Comment by tazjin** [comment]
In fact, a followup to that password comment: If the certificate is in PKCS12 format it will contain the key and require a passphrase, in that case only `Easy::ssl_cert` and `Easy::key_password` should be called (not `Easy::ssl_key`).

I've experimented a bit with making a higher level wrapper over the `curl` crate myself now and I can't find any other way to handle this than splitting up the `ssl_cert`, `ssl_key` and `key_password` functions in a similar way as cURL itself does. It may actually be easiest to it that way and assume that users setting these options "know what they're doing" ... ?

**Comment by sagebind** [comment]
Interesting, my knowledge of certificates isn't very comprehensive. I bet we could encode these rules into the type system with an enum:

```rust
pub enum ClientCertificate {
    PEM {
        path: PathBuf,
        private_key: Option<PrivateKey>, // unlikely to be None, but possible
    },
    DER {
        path: PathBuf,
        private_key: Option<PrivateKey>, // unlikely to be None, but possible
    },
    P12 {
        path: PathBuf,
        password: String,
    },
}

pub enum PrivateKey {
    PEM {
        path: PathBuf,
        password: Option<String>,
    },
    DER {
        path: PathBuf,
        password: Option<String>,
    },
}
```

What do you think?

**Comment by tazjin** [comment]
Nice, that seems close! I'm not sure if a PKCS12 key pair *can* omit password protection, will try to look that up later (unless you get to it first!).

Some rabbitholing:

1. There are some special oddities on OS X, where the path may in fact not be a path but simply the name of a Keychain item - but that's already weird in the upstream `curl` crate so it probably doesn't matter much.

2. Would anyone ever have a differently encoded private and public key? I guess it's technically possible and humans will do anything. However, take note of [this comment](https://docs.rs/curl/0.4.19/curl/easy/struct.Easy2.html#method.ssl_key_type) which says:

    > "DER" format key file currently does not work because of a bug in OpenSSL.

    Which may be interesting to propagate, but since that doesn't link anywhere I'm also not sure it's even true anymore.

3. I also agree with not supporting the `ENG` format here. People who use that sort of thing to interface with crypto hardware and the like can probably just go ahead and use `curl` directly. I did experiment with [exposing the `Easy` struct to the user](https://docs.rs/crimp/0.1.0/crimp/struct.Request.html#method.with_handle) in my experimental cURL-API crate, if you wanted to support more advanced use-cases that sort of thing would be one option.

It may make sense to look at the config of other software with certificate configuration (e.g. `nginx`) to see how they handle all these cases. It's nice to have a reasonably familiar interface when dealing with certificate-related APIs.

**Comment by sagebind** [comment]
Some initial thoughts without doing any research:

>     1. There are some special oddities on OS X, where the path may in fact not be a path but simply the name of a Keychain item

I think it may be better to just pretend that isn't a thing.

>     2. Would anyone ever have a differently encoded private and public key?

It's legal, so the types should probably allow it. Even though it is pretty weird.

>     "DER" format key file currently does not work because of a bug in OpenSSL.

I'd be curious to find out what _does not work_ means. If OpenSSL simply returns an error, then that's fine if we just propagate the error. If it means subtle bugs will happen, then that's a different story. Either way, slapping a warning on it in the docs is likely sufficient.

>     3. I also agree with not supporting the `ENG` format here.

Yeah, reading the curl docs it looked like a really hard thing to check and use safely, so I decided to omit it.

If advanced uses of curl are needed, my current strategy is to just recommend using curl directly. Right now none of the curl crate types are exposed in cHTTP's public API and I'd like to keep it that way if I can.

**Comment by tazjin** [comment]
> I'd be curious to find out what _does not work_ means.

It's relatively safe to assume that whatever happens is a pretty disruptive change, if the TLS handshake can't succeed there won't be any data transferring going on. Makes it very likely that it will just come up as some variant of `curl::Error`.

> Right now none of the curl crate types are exposed in cHTTP's public API and I'd like to keep it that way if I can

:+1: 

On the PKCS12 thing: It seems to be possible (albeit somewhat unusual) to remove password protection from a PKCS12 file, so I think that `password` field should be an `Option`. Should be good to go after that, if somebody finds a use-case that breaks this setup they should probably file an issue or rethink what they're doing :grinning:

**Pull request merged**